### PR TITLE
codeintel: Do not error on missing HEAD while auto-indexing

### PR DESCRIPTION
**Pull request by efritz**
Fixes https://github.com/sourcegraph/sourcegraph/issues/22357.

### CODE REVIEW

File: enterprise/internal/codeintel/gitserver/client.go [hunk 1]
```diff
@@ -51,14 +51,25 @@ func (c *Client) CommitExists(ctx context.Context, repositoryID int, commit stri
...
 	defer endObservation(1, observation.Args{})
 
-	return c.execGitCommand(ctx, repositoryID, "rev-parse", "HEAD")
+	revision, err := c.execGitCommand(ctx, repositoryID, "rev-parse", "HEAD")
+	if err != nil {
+		if isRevisionNotFound(err) {
+			return "", false, nil
+		}
+
+		return "", true, err
```
Comment (efritz) [inline enterprise/internal/codeintel/gitserver/client.go:69]:
```suggestion
		return "", false, err
```

File: enterprise/internal/codeintel/gitserver/client.go [hunk 2]
```diff
@@ -51,14 +51,25 @@ func (c *Client) CommitExists(ctx context.Context, repositoryID int, commit stri
...
 	ctx, endObservation := c.operations.head.With(ctx, &err, observation.Args{LogFields: []log.Field{
 		log.Int("repositoryID", repositoryID),
 	}})
 	defer endObservation(1, observation.Args{})
 
-	return c.execGitCommand(ctx, repositoryID, "rev-parse", "HEAD")
+	revision, err := c.execGitCommand(ctx, repositoryID, "rev-parse", "HEAD")
+	if err != nil {
+		if isRevisionNotFound(err) {
+			return "", false, nil
```
Comment (efritz) [inline enterprise/internal/codeintel/gitserver/client.go:66]:
```suggestion
			return "", false, nil
```

File: enterprise/cmd/worker/internal/codeintel/commitgraph/updater.go [hunk 3]
```diff
@@ -108,8 +108,8 @@ func (u *Updater) update(ctx context.Context, repositoryID, dirtyToken int) (err
 	}
 	traceLog(log.Int("numCommitGraphKeys", len(commitGraph.Order())))
 
-	tipCommit, err := u.gitserverClient.Head(ctx, repositoryID)
-	if err != nil {
+	tipCommit, ok, err := u.gitserverClient.Head(ctx, repositoryID)
+	if err != nil || !ok {
 		return errors.Wrap(err, "gitserver.Head")
```
Comment (Strum355) [inline enterprise/cmd/worker/internal/codeintel/commitgraph/updater.go:113]: When `!ok`, we're wrapping a nil error. Is this what we want?

File: enterprise/internal/codeintel/gitserver/client.go [hunk 4]
```diff
@@ -426,3 +437,13 @@ func (c *Client) repositoryIDToRepo(ctx context.Context, repositoryID int) (api.
...
+
+func isRevisionNotFound(err error) bool {
+	for ex := err; ex != nil; ex = errors.Unwrap(ex) {
+		if gitserver.IsRevisionNotFound(ex) {
+			return true
+		}
+	}
+
+	return false
+}
```
Comment (Strum355) [inline enterprise/internal/codeintel/gitserver/client.go:449]:
```suggestion
func isRevisionNotFound(err error) bool {
	return errors.Is(err, &gitserver.RevisionNotFoundError{})
}
```

Or just inline that.

File: enterprise/cmd/worker/internal/codeintel/commitgraph/updater.go (hunk 3 above)
Comment (efritz) [reply enterprise/cmd/worker/internal/codeintel/commitgraph/updater.go:113]: `errors.Wrap(nil) = nil`

Comment (Strum355) [reply enterprise/cmd/worker/internal/codeintel/commitgraph/updater.go:113]: Funky but ok

Comment (efritz) [reply enterprise/cmd/worker/internal/codeintel/commitgraph/updater.go:113]: Seems idiomatic with other things we do (basestore.CloseRows, observability stuff, etc)

**Review by tjdevries — APPROVED** [review]
LGTM

**Pull request merged**